src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,515 @@ func (s *commitLogSource) ReadData(
+							xlog.NewField("filepaths", mostRecentSnapshot.AbsoluteFilepaths),
+						).
+						Error("error resolving snapshot time for snapshot file")
+
+					// If we couldn't determine the snapshot time for the snapshot file, then rely
+					// on the defer to fallback to using the block start time.
+					return
+				}
+
+				mostRecentSnapshot = mostRecentSnapshotVolume
+			}()
+		}
+	}
+
+	return mostRecentSnapshotsByBlockShard
+}
+
+func (s *commitLogSource) minimumMostRecentSnapshotTimeByBlock(
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	mostRecentSnapshotByBlockShard map[xtime.UnixNano]map[uint32]fs.FileSetFile,
+) map[xtime.UnixNano]time.Time {
+	minimumMostRecentSnapshotTimeByBlock := map[xtime.UnixNano]time.Time{}
+	for blockStart, mostRecentSnapshotsByShard := range mostRecentSnapshotByBlockShard {
+
+		var minMostRecentSnapshot time.Time
+		for shard, mostRecentSnapshotForShard := range mostRecentSnapshotsByShard {
+			blockRange := xtime.Range{Start: blockStart.ToTime(), End: blockStart.ToTime().Add(blockSize)}
+			if !shardsTimeRanges[shard].Overlaps(blockRange) {
+				// In order for a minimum most recent snapshot to be valid, it needs to be for a block that
+				// we actually need to bootstrap for that shard. This check ensures that our algorithm doesn't
+				// do any extra work, even if we're bootstrapping different blocks for various shards.
+				continue
+			}
+
+			if mostRecentSnapshotForShard.CachedSnapshotTime.Before(minMostRecentSnapshot) || minMostRecentSnapshot.IsZero() {
+				minMostRecentSnapshot = mostRecentSnapshotForShard.CachedSnapshotTime
+			}
+		}
+
+		if minMostRecentSnapshot.IsZero() {
+			// If we didn't find a minimum most recent snapshot time for this blockStart, just use the
+			// blockStart as the minimum since we'll need to read the entire commit log in this case.
+			minMostRecentSnapshot = blockStart.ToTime()
+		}
+
+		minimumMostRecentSnapshotTimeByBlock[blockStart] = minMostRecentSnapshot
+	}
+
+	return minimumMostRecentSnapshotTimeByBlock
+}
+
+func (s *commitLogSource) bootstrapShardSnapshots(
+	nsID ident.ID,
+	shard uint32,
+	metadataOnly bool,
+	shardTimeRanges xtime.Ranges,
+	blockSize time.Duration,
+	snapshotFiles fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshotByBlockShard map[xtime.UnixNano]map[uint32]fs.FileSetFile,
+) (result.ShardResult, error) {
+	var (
+		shardResult    result.ShardResult
+		allSeriesSoFar *result.Map
+		rangeIter      = shardTimeRanges.Iter()
+		err            error
+	)
+
+	for rangeIter.Next() {
+
+		var (
+			currRange             = rangeIter.Value()
+			currRangeDuration     = currRange.End.Sub(currRange.Start)
+			isMultipleOfBlockSize = currRangeDuration%blockSize == 0
+		)
+
+		if !isMultipleOfBlockSize {
+			return nil, fmt.Errorf(
+				"received bootstrap range that is not multiple of blockSize, blockSize: %d, start: %d, end: %d",
+				blockSize, currRange.End.Unix(), currRange.Start.Unix(),
```

Comment:
Perhaps best to log dates as string like we do elsewhere?

Reply:
done